dropshot/examples/file_server.rs

```diff
@@ -0,0 +1,213 @@
+// Copyright 2021 Oxide Computer Company
+
+use dropshot::ApiDescription;
+use dropshot::ConfigLogging;
+use dropshot::ConfigLoggingLevel;
+use dropshot::HttpError;
+use dropshot::HttpServerStarter;
+use dropshot::RequestContext;
+use dropshot::{endpoint, Path};
+use http::{Response, StatusCode};
+use hyper::Body;
+use schemars::JsonSchema;
+use serde::Deserialize;
+use std::path::PathBuf;
+use std::sync::Arc;
+
+/**
+ * Our context is simply the root of the directory we want to serve.
+ */
+struct FileServerContext {
+    base: PathBuf,
+}
+
+#[tokio::main]
+async fn main() -> Result<(), String> {
+    /*
+     * We must specify a configuration with a bind address.  We'll use 127.0.0.1
+     * since it's available and won't expose this server outside the host.  We
+     * request port 0, which allows the operating system to pick any available
+     * port.
+     */
+    let config_dropshot = Default::default();
+
+    /*
+     * For simplicity, we'll configure an "info"-level logger that writes to
+     * stderr assuming that it's a terminal.
+     */
+    let config_logging = ConfigLogging::StderrTerminal {
+        level: ConfigLoggingLevel::Info,
+    };
+    let log = config_logging
+        .to_logger("example-basic")
+        .map_err(|error| format!("failed to create logger: {}", error))?;
+
+    /*
+     * Build a description of the API -- in this case it's not much of an API!.
+     */
+    let mut api = ApiDescription::new();
+    api.register(static_content).unwrap();
+
+    /*
+     * Specify the directory we want to serve.
+     */
+    let context = FileServerContext {
+        base: PathBuf::from("."),
+    };
+
+    /*
+     * Set up the server.
+     */
+    let server = HttpServerStarter::new(&config_dropshot, api, context, &log)
+        .map_err(|error| format!("failed to create server: {}", error))?
+        .start();
+
+    /*
+     * Wait for the server to stop.  Note that there's not any code to shut down
+     * this server, so we should never get past this point.
+     */
+    server.await
+}
+
+/**
+ * Dropshot deserializes the input path into this Vec.
+ */
+#[derive(Deserialize, JsonSchema)]
+struct AllPath {
+    path: Vec<String>,
+}
+
+/**
+ * Serve files from the specified root path.
+ */
+#[endpoint {
+    method = GET,
+
+    /*
+     * Match literally every path including the empty path.
+     */
+    path = "/{path:.*}",
```

Comment:
I'm double checking: I think I saw a PR about these sorts of paths that included questions like can `..` be a path component? Or am I thinking of something else?

Reply:
You may well have and the answer is that we don't allow so-called "dot-segments" to be present. See https://github.com/oxidecomputer/dropshot/blob/main/dropshot/src/router.rs#L769

Reply:
Yup, that's the one!